File: src/main/java/com/mkreidl/ephemeris/geometry/ClassicalOrbitalElements.java
package com.mkreidl.ephemeris.geometry;

import static java.lang.Math.*;

public class ClassicalOrbitalElements extends OrbitalElements
{
    public double node;
    public double incl;
    public double periapsis;
    public double axis;
    public double exc;
    public double meanAnom;

    public ClassicalOrbitalElements set( ClassicalOrbitalElements orig )
    {
        this.node = orig.node;
        this.incl = orig.incl;
        this.periapsis = orig.periapsis;
        this.axis = orig.axis;
        this.exc = orig.exc;
        this.meanAnom = orig.meanAnom;
        return this;
    }

    public ClassicalOrbitalElements set( double[] values )
    {
        this.node = values[0];
        this.incl = values[1];
        this.periapsis = values[2];
        this.axis = values[3];
        this.exc = values[4];
        this.meanAnom = values[5];
        return this;
    }

    public double meanLongitude()
    {
        return node + periapsis + meanAnom;
    }

    /**
     * Solve Kepler's equation with Newton method
     *
     * @param eps Required precision
     * @return
     */
    public double eccentricAnomaly( double eps )
    {
        double e0;
        double excentricAnom = meanAnom;
        // Solve Kepler's equation E - e*sin(E) = M by Newton iteration
        do
        {
            e0 = excentricAnom;
            excentricAnom = e0 - ( e0 - exc * sin( e0 ) - meanAnom ) / ( 1.0 - exc * cos( e0 ) );
        } while ( abs( excentricAnom - e0 ) > eps );
        return excentricAnom;
    }

    public void toVSOP87OrbitalElements( VSOP87OrbitalElements output )
    {
        output.a = axis;
        output.l = meanLongitude();
        output.h = exc * sin( node + periapsis );
        output.k = exc * cos( node + periapsis );
        output.p = sin( 0.5 * incl ) * sin( node );
        output.q = sin( 0.5 * incl ) * cos( node );
    }

    public void computePosition()
    {
        double eccentricAnom = eccentricAnomaly( 1e-15 );

        // Cartesian coordinates in the orbital plane,
        // The ascending node defines the x-axis
        double xv = axis * ( cos( eccentricAnom ) - exc );
        double yv = axis * ( sqrt( 1.0 - exc * exc ) * sin( eccentricAnom ) );

        // Cartesian coordinates after rotation to account for periapsis
        double xw = xv * cos( periapsis ) - yv * sin( periapsis );
        double yw = xv * sin( periapsis ) + yv * cos( periapsis );

        eclipticalCartesian.x = cos( node ) * xw - sin( node ) * yw * cos( incl );
        eclipticalCartesian.y = sin( node ) * xw + cos( node ) * yw * cos( incl );
        eclipticalCartesian.z = yw * sin( incl );

        eclipticalCartesian.transform( eclipticalSpherical );
    }

    public void add( ClassicalOrbitalElements summand )
    {
        this.node += summand.node;
        this.incl += summand.incl;
        this.periapsis += summand.periapsis;
        this.axis += summand.axis;
        this.exc += summand.exc;
        this.meanAnom += summand.meanAnom;
        standardize();
    }

    public void times( double x )
    {
        this.node *= x;
        this.incl *= x;
        this.periapsis *= x;
        this.axis *= x;
        this.exc *= x;
        this.meanAnom *= x;
        standardize();
    }

    private void standardize()
    {
        this.node = Angle.standardize( this.node );
        this.incl = Angle.standardize( this.incl );
        this.periapsis = Angle.standardize( this.periapsis );
        this.meanAnom = Angle.standardize( this.meanAnom );
    }

    @Override
    public String toString()
    {
        return "ClassicalOrbitalElements [node="
                + node
                + ", incl="
                + incl
                + ", periapsis="
                + periapsis
                + ", axis="
                + axis
                + ", exc="
                + exc
                + ", meanAnom="
                + meanAnom
                + "]";
    }

}

File: src/main/java/com/mkreidl/ephemeris/solarsystem/ClassicalOrbitalElements.kt
package com.mkreidl.ephemeris.solarsystem

import com.mkreidl.ephemeris.solarsystem.vsop87.Vsop87OrbitalElements
import com.mkreidl.math.Angle
import com.mkreidl.math.Coordinates
import com.mkreidl.math.PhaseCartesian
import com.mkreidl.math.Vector3

/**
 * N = longitude of the ascending node
 * i = inclination to the ecliptic (plane of the Earth's orbit)
 * w = argument of perihelion
 * a = semi-major axis
 * e = eccentricity (0=circle, 0-1=ellipse, 1=parabola)
 * M = mean anomaly (0 at perihelion; increases uniformly with time)
 */
data class ClassicalOrbitalElements(
        val node: Double,
        val inclination: Double,
        val periapsis: Double,
        val axis: Double,
        val excentricity: Double,
        val meanAnomaly: Double
) : OrbitalElements {

    override val classical get() = this

    override val vsop87 by lazy {
        Vsop87OrbitalElements(
                a = axis,
                l = meanLongitude,
                h = excentricity * Math.sin(node + periapsis),
                k = excentricity * Math.cos(node + periapsis),
                p = Math.sin(0.5 * inclination) * Math.sin(node),
                q = Math.sin(0.5 * inclination) * Math.cos(node)
        )
    }

    override fun computePhase() = PhaseCartesian(computePosition().cartesian, Vector3.ZERO)

    override fun computePosition(): Coordinates {
        val eccentricAnomaly = computeEccentricAnomaly(1e-13, 10)
        // Cartesian coordinates in the orbital plane,
        // The ascending node defines the x-axis
        val xv = axis * (Math.cos(eccentricAnomaly) - excentricity)
        val yv = axis * (Math.sqrt(1.0 - excentricity * excentricity) * Math.sin(eccentricAnomaly))
        // Cartesian coordinates after rotation to account for periapsis
        val xw = xv * Math.cos(periapsis) - yv * Math.sin(periapsis)
        val yw = xv * Math.sin(periapsis) + yv * Math.cos(periapsis)
        return Vector3(
                x = Math.cos(node) * xw - Math.sin(node) * yw * Math.cos(inclination),
                y = Math.sin(node) * xw + Math.cos(node) * yw * Math.cos(inclination),
                z = yw * Math.sin(inclination)
        )
    }

    val meanLongitude get() = node + periapsis + meanAnomaly

    /**
     * Solve Kepler's equation with Newton method
     *
     * @param eps Required precision
     * @return
     */
    fun computeEccentricAnomaly(eps: Double, maxIter: Int = Int.MAX_VALUE): Double {
        var e0: Double
        var excentricAnomaly = meanAnomaly
        var iter = 0
        // Solve Kepler's equation E - e*Math.sin(E) = M by Newton iteration
        do {
            e0 = excentricAnomaly
            excentricAnomaly = e0 - (e0 - excentricity * Math.sin(e0) - meanAnomaly) / (1.0 - excentricity * Math.cos(e0))
            println(excentricAnomaly)
            ++iter
        } while (iter < maxIter && Math.abs(excentricAnomaly - e0) > eps)
        return excentricAnomaly
    }

    fun reduce() = copy(
            node = Angle.reduce(node),
            inclination = Angle.reduce(inclination),
            periapsis = Angle.reduce(periapsis),
            meanAnomaly = Angle.reduce(meanAnomaly)
    )
}
